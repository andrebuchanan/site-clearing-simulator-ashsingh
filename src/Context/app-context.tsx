import React, { ReactNode, useMemo, useState } from 'react'

type Session = {
  init: boolean
  quit: boolean
}

interface AppContextInterface {
  fuelConsumed: number[]
  itemizedCosts: number[]
  loggedCommands: string[]
  sessionState: Session
  siteData: string[][]
  logCommand: (command: string) => void
  saveSiteData: (data: string[][]) => void
  updateCost: (price: number) => void
  updateFuel: (quantity: number) => void
  updateSession: (session: Session) => void
}

export const AppContext = React.createContext<AppContextInterface>({
  fuelConsumed: [],
  itemizedCosts: [],
  loggedCommands: [],
  sessionState: { init: true, quit: false },
  siteData: [],
  logCommand: () => undefined,
  saveSiteData: () => undefined,
  updateCost: () => undefined,
  updateFuel: () => undefined,
  updateSession: () => undefined
})

export const AppContextProvider = ({ children }: { children: ReactNode }) => {
  const [commands, setCommands] = useState<string[]>([])
  const [cost, setCost] = useState<number[]>([])
  const [fuel, setFuel] = useState<number[]>([])
  const [session, setSession] = useState<Session>({ init: true, quit: false })
  const [siteInput, setSiteInput] = useState<string[][]>([])

  const logCommandHandler = (command: string) => {
    setCommands((prevState) => [...prevState, command])
  }

  const saveSiteDataHandler = (data: string[][]) => {
    setSiteInput(data)
  }

  const updateCostHandler = (price: number) => {
    setCost((prevState) => [...prevState, price])
  }

  const updateFuelHandler = (quantity: number) => {
    setFuel((prevState) => [...prevState, quantity])
  }

  const updateSessionHandler = (sessionProp: Session) => {
    console.log('session', sessionProp)
    if (sessionProp.quit) console.log('quit')
    setSession(sessionProp)
  }

  const contextValue = useMemo(
    () => ({
      fuelConsumed: fuel,
      itemizedCosts: cost,
      loggedCommands: commands,
      sessionState: session,
      siteData: siteInput,
      logCommand: logCommandHandler,
      saveSiteData: saveSiteDataHandler,
      updateCost: updateCostHandler,
      updateFuel: updateFuelHandler,
      updateSession: updateSessionHandler
    }),
    [commands, cost, fuel, session, siteInput]
  )

  return (
    <AppContext.Provider value={contextValue}>{children}</AppContext.Provider>
  )
}

export default AppContextProvider
