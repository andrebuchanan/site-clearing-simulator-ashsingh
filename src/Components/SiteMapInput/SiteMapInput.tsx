import React, { useContext, useState } from 'react'
import { useNavigate } from 'react-router-dom'

import Box from '@mui/material/Box'
import Button from '@mui/material/Button'
import Grid from '@mui/material/Grid'
import { styled } from '@mui/material/styles'
import Typography from '@mui/material/Typography'

import { AppContext } from '../../Context'

import { InvalidFileDialog } from '..'

const SiteMapInput = () => {
  const { saveSiteData } = useContext(AppContext)
  const [fileInput, setFileInput] = useState<string[][]>([])
  const [invalid, setInvalid] = useState(false)
  const navigate = useNavigate()

  React.useEffect(() => {
    if (fileInput.length) {
      saveSiteData(fileInput)
      navigate('/sitemap')
    }
  }, [fileInput])

  const Input = styled('input')({
    display: 'none'
  })

  const handleFileInput = async (
    event: React.ChangeEvent<HTMLInputElement>
  ) => {
    event.preventDefault()
    const files = event.target.files || []
    const file = files[0]
    // return if not text file
    if (!file || file.type !== 'text/plain') return setInvalid(true)

    const reader = new FileReader()

    reader.onload = async (e) => {
      const text = e.target?.result
      if (!text) return setInvalid(true)
      // create 1D Array from text
      const textArr = text?.toString().split('\n')
      if (!textArr) return setInvalid(true)
      // Create 2D Array
      const matrix = textArr?.map((row) => row.split(''))
      if (!matrix) return setInvalid(true)
      // validate 2D Array
      const validMatrix = matrix?.every(
        (row) => row.length === matrix[0].length
      )
      if (!validMatrix) return setInvalid(true)

      return setFileInput(matrix)
    }
    return reader.readAsText(file)
  }

  return (
    <Box sx={{ flexGrow: 1 }}>
      <Grid container spacing={2}>
        <Grid item xs={4}>
          <Typography variant="h6" component="div" gutterBottom>
            Upload a sitemap to begin session.
          </Typography>
        </Grid>
        <Grid item xs={4}>
          <label htmlFor="contained-button-file">
            <Input
              id="contained-button-file"
              data-testid="contained-button-file"
              type="file"
              onChange={handleFileInput}
            />
            <Button variant="contained" component="span">
              Upload
            </Button>
          </label>
        </Grid>
      </Grid>
      <InvalidFileDialog
        isOpen={invalid}
        handleClick={() => setInvalid(false)}
      />
    </Box>
  )
}

export default SiteMapInput
