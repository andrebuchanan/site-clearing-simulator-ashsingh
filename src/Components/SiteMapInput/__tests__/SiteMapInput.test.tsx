import React from 'react'
import { BrowserRouter } from 'react-router-dom'
import { fireEvent, render, screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import { AppContext } from '../../../Context'
import SiteMapInput from '../SiteMapInput'

const contextValue = {
  saveSiteData: jest.fn()
}

test('renders SiteMapInput', () => {
  const { asFragment } = render(
    <BrowserRouter>
      <AppContext.Provider value={contextValue}>
        <SiteMapInput />
      </AppContext.Provider>
    </BrowserRouter>
  )

  expect(asFragment()).toMatchSnapshot()
})

test('uploads text file', async () => {
  const file = new File(['oootoo\noootoo'], 'input.txt', { type: 'text/plain' })
  render(
    <BrowserRouter>
      <AppContext.Provider value={contextValue}>
        <SiteMapInput />
      </AppContext.Provider>
    </BrowserRouter>
  )
  const uploadButton = screen.getByTestId('contained-button-file')

  await waitFor(() => userEvent.upload(uploadButton, file))

  const textFile = document.getElementById('contained-button-file')

  expect(textFile.files[0].name).toBe('input.txt')
  expect(textFile.files.length).toBe(1)
})

test('saves file data', async () => {
  const file = new File(['oootoo\noootoo'], 'input.txt', { type: 'text/plain' })
  const mockFn = jest.fn()
  render(
    <BrowserRouter>
      <AppContext.Provider value={{ saveSiteData: mockFn }}>
        <SiteMapInput />
      </AppContext.Provider>
    </BrowserRouter>
  )
  const uploadButton = screen.getByTestId('contained-button-file')

  await waitFor(() => userEvent.upload(uploadButton, file))

  const textFile = document.getElementById('contained-button-file')

  expect(textFile.files[0].name).toBe('input.txt')
  expect(textFile.files.length).toBe(1)
  await tick()
  // expect(mockFn).toHaveBeenCalled()
})

function tick() {
  return new Promise((resolve) => {
    setTimeout(resolve, 0)
  })
}
