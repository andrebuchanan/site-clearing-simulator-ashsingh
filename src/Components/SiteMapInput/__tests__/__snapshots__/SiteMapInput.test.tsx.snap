// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders SiteMapInput 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-i9gxme"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
      >
        <label
          for="contained-button-file"
        >
          <input
            class="css-1hyfx7x"
            data-testid="contained-button-file"
            id="contained-button-file"
            type="file"
          />
          <span
            class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
            role="button"
            tabindex="0"
          >
            Upload
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </span>
        </label>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
